### DISCUSSION TITLE
Adding decoder for Shelly Bluetooth Humidity and Temp (SBHT-003C) Sensor

### DISCUSSION THREAD

**Discussion by chrestme** [theengs/decoder#553] [New devices/things integration]
I'm trying to add a decoder for the Shelly Bluetooth Humidity and Temp Sensor. I've followed the tutorial from [theengs.io](https://decoder.theengs.io/participate/adding-decoders.html) by 

1. cloned the repo 
2. copied the `src` directory under the `python` directory
3. Created `SBHT_003C_json.h` under `python/src/devices/`
4. Ran `scripts/check_decoder.py` to fix up the JSON string.
5. Modified `devices.h` and `decoder.h` to include references to SBHT_003C
6. Ran `pip install .` in the python directory to build/rebuild the project.
7. Ran `ScanAndDecode.py` to collect a sample of the SBHT advertising packet
8. Copied the JSON string from the `ScanAndDecode.py` output into `test/python_ble/decoder_test.py`
9. Ran `decoder_test.py`

The result of decoder_test.py returns None for the call to decodeBLE (dble). 

This is the string I got from ScanandDecode.py
`{"servicedatauuid": "fcd2", "servicedata": "4400ae01642e3845fa00", "manufacturerdata": "a90b0111000b03000af53562b6c67c", "name": "SBHT-003C", "id": "7C:C6:B6:62:35:F5", "rssi": -88}`

At first I had a full parser defined in my SBHT_003C_json.h (using SBT_002C_json.h as an example), but when that didn't work I started pairing down the matching conditions until I just had one. So, now it just looks like this:

 ```
{ 
    "brand":"Shelly",
    "model":"ShellyBLU H&T",
    "model_id":"SBHT-003C",
    "tag":"1106",
    "condition":["name", "index", 0, "SBHT-003C"],
    "properties":{}
 }
```

I've tried different simple conditions, but none of them return anything for decodeBLE. 
`"condition":["uuid", "index", 0, "fcd2"]`
`"condition":["servicedata", "=", 20, "index", 0, "44"]`

I checked to make sure these changes were getting picked up by running strings on the compiled .so file, and grepped for SBHT-003C, and got the expected result.

```
$ strings ~/theengs/lib/python3.10/site-packages/TheengsDecoder/_decoder.cpython-310-x86_64-linux-gnu.so | grep SBHT`
{"brand":"Shelly","model":"ShellyBLU H&T","model_id":"SBHT-003C","tag":"1106","condition":["name","index",0,"SBHT-003C"],"properties":{}}
_SBHT_003C_json
_SBHT_003C_json_props
```

Thanks in advance for any help.

**Comment by DigiH** [comment]
Hi @chrestme,

Great that you went through the documentation and started your decoder for the Shelly Bluetooth Humidity and Temp (SBHT-003C).

Any chances of having your forked repo here on GitHub, so that looking at the new branch should show what might be the issue would be a lot easier?

> but when that didn't work I started pairing down the matching conditions until I just had one. So, now it just looks like this:

If there are no decoded properties at all there won't be any decoded publishing, so this could be your current case, but looking at the whole changes you made really is best.

**Reply by DigiH** [reply]
Looking at the details for the SBHT-003C, I have come up with the following decoder and using your above and @GeoChallenger's sample broadcast data as test cases. 

I've also added the two possible button presses which are awkwardly injected before the temperature , also making the servicedata longer - **I assume so, some sample data from the actual device during one of the two possible button press/hold would be needed to verify this.**

While the test cases run fine, could you please test this with your actual devices? 

https://github.com/theengs/decoder/pull/554

The next step would also be to get some encrypted sample data, so that the encrypted decoder could also be added to this model as with the other Shelly devices.

Thanks

**Reply by chrestme** [reply]
Thanks for taking this up so quickly. I pulled it down and tested and it works! Thanks for your help with this! 

Output from ScanAndDecode.py:
```
7C:C6:B6:62:35:F5:AdvertisementData(local_name='SBHT-003C', manufacturer_data={2985: b'\x01\x11\x00\x0b\x03\x00\n\xf55b\xb6\xc6|'}, service_data={'0000fcd2-0000-1000-8000-00805f9b34fb': b'D\x00\xbe\x01d./E\xf5\x00'}, rssi=-90)

data sent to decoder:  {"servicedatauuid": "fcd2", "servicedata": "4400be01642e2f45f500", "manufacturerdata": "a90b0111000b03000af53562b6c67c", "name": "SBHT-003C", "id": "7C:C6:B6:62:35:F5", "rssi": -90}

TheengsDecoder found device: {"servicedatauuid":"fcd2","servicedata":"4400be01642e2f45f500","manufacturerdata":"a90b0111000b03000af53562b6c67c","name":"SBHT-003C","id":"7C:C6:B6:62:35:F5","rssi":-90,"brand":"Shelly","model":"ShellyBLU H&T","model_id":"SBHT-003C","type":"THB","acts":true,"cont":true,"packet":190,"batt":100,"hum":47,"tempc":24.5,"tempf":76.1,"mac":"7C:C6:B6:62:35:F5"}
7C:C6:B6:62:35:F5: SBHT-003C
```

Looking at your commits, it looks pretty much exactly like what I had, so I'm not sure what the issue was, but glad it works now.

**Reply by DigiH** [reply]
Thanks for the verification @chrestme 

Could you also do this for the two possible 
    **1 - Single short click
    11 - Button hold**
just to see if my assumption of the button code being injected before the temperature is correct, or if it might be added to the end of the servicedata?

And would you possibly be able to also test the encryption functionality with [**Theengs Gateway**](https://github.com/theengs/gateway)?

https://gateway.theengs.io/use/use.html#reading-encrypted-advertisements

> Looking at your commits, it looks pretty much exactly like what I had, so I'm not sure what the issue was

Hard to say without having seen your earlier code, but a diff of the two might show up some very small difference ;)

**Comment by chrestme** [comment]
I'll have to test it tonight when I get home and have access to the device,
but from my previous testing you are correct, the button press data gets
injected before the tempc data.

**Reply by DigiH** [reply]
That would be great.

Some raw encrypted servicedata first so I can finish the encryption decoder as well, and once added to the existing branch/PR it will also be testable with Theengs Gateway.

Thanks

**Reply by chrestme** [reply]
I was having trouble with my Shelly app last night so I wasn't able to enable encryption. (Turns out I had an old version of the app?) I think I have the correct version now, so I'll give it another try.

In the mean time, I was able to collect some button press servicedata samples.

**Single Press**
```
data sent to decoder:  {"servicedatauuid": "fcd2", "servicedata": "44003001642e333a01450a01", "manufacturerdata": "a90b0111000b03000af53562b6c67c", "name": "SBHT-003C", "id": "7C:C6:B6:62:35:F5", "rssi": -89}

TheengsDecoder found device: {"servicedatauuid":"fcd2","servicedata":"44003001642e333a01450a01","manufacturerdata":"a90b0111000b03000af53562b6c67c","name":"SBHT-003C","id":"7C:C6:B6:62:35:F5","rssi":-89,"brand":"Shelly","model":"ShellyBLU H&T","model_id":"SBHT-003C","type":"THB","acts":true,"cont":true,"packet":48,"batt":100,"hum":51,"button":1,"tempc":26.6,"tempf":79.88,"mac":"7C:C6:B6:62:35:F5"}
```

**Long Press**
```
data sent to decoder:  {"servicedatauuid": "fcd2", "servicedata": "44003301642e313afe450f01", "manufacturerdata": "a90b0111000b03000af53562b6c67c", "name": "SBHT-003C", "id": "7C:C6:B6:62:35:F5", "rssi": -92}

TheengsDecoder found device: {"servicedatauuid":"fcd2","servicedata":"44003301642e313afe450f01","manufacturerdata":"a90b0111000b03000af53562b6c67c","name":"SBHT-003C","id":"7C:C6:B6:62:35:F5","rssi":-92,"brand":"Shelly","model":"ShellyBLU H&T","model_id":"SBHT-003C","type":"THB","acts":true,"cont":true,"packet":51,"batt":100,"hum":49,"button":11,"tempc":27.1,"tempf":80.78,"mac":"7C:C6:B6:62:35:F5"}
```

**Reply by DigiH** [reply]
@chrestme Thanks for the verification of the button presses, and the confirmation that the button octets are injected in between the humidity and temperature ones.

I hope you can get the encryption working with the new app version. Just a sample of the servicedata/manufacturerdata would be required in the normal temp/hum/batt mode, and for a button press to see if the encrypted button press advertising data is also longer, with the button hold likely being the easiest for that undecoded test. This would allow me to also create the decryption decoder for the SBHT-003C

**Reply by chrestme** [reply]
Sorry for the delay. I was able to set up encryption on the SBHT sensor and got the following packets:

Edit - Realized I forgot the pin: 123456

**Regular Advertisement**
```
{"manufacturerdata": "a90b0111000b03000af53562b6c67c", "name": "SBHT-003C", "id": "7C:C6:B6:62:35:F5", "rssi": -61, "servicedatauuid": "fcd2", "servicedata": "45e1a236ace4b40861674bfbd3668c080c2e", "mfr": "Allterco Robotics ltd"}
```

**Single Press**
```
{"manufacturerdata": "a90b0111000b03000af53562b6c67c", "name": "SBHT-003C", "id": "7C:C6:B6:62:35:F5", "rssi": -66, "servicedatauuid": "fcd2", "servicedata": "45c85de27369ce3f84750015aae9d466b801349b", "mfr": "Allterco Robotics ltd"}
```

**Long Press**
```
{"manufacturerdata": "a90b0111000b03000af53562b6c67c", "name": "SBHT-003C", "id": "7C:C6:B6:62:35:F5", "rssi": -68, "servicedatauuid": "fcd2", "servicedata": "453a25e3518e72c92f6dfef8e9e9d46696b53b8a", "mfr": "Allterco Robotics ltd"}
```

**Reply by DigiH** [reply]
Great, thanks a lot @chrestme!

While I can't really test it out here without an actual device, everything looks as expected, with the encrypted data also being two octets longer than the pure temp,hum,batt data, and the added test cases matching the expected results :)

I think the PR is ready for merging, unless you have any further comments.

**Reply by andrewplas** [reply]
Thanks @DigiH and @chrestme!

What is the likelihood this can get merged into OpenMQTTGateway?  I have several OMG BLE gateways handling Shelly SBDW002C sensors and Switchbot temp/hum meters without issue, and I'd like to start using the Shelly SBHT-003C as well.

**Reply by DigiH** [reply]
Hi @andrewplas 

How do you upload OpenMQTTGateway to your gateways?

If you are using environments with PlatformIO or gitpod.io, you can replace the Theengs Decoder reference in the platformio.ini file from 
`decoder = https://github.com/theengs/decoder.git#v1.6.4`
to
`decoder = https://github.com/theengs/decoder.git#development`

or you could install a pre-built binary from the web upload page of the current development version of OpenMQTTGateway. Just be aware that it might have an AirPods test decoder included in the next days - the current built with **SHA:952a65** is fine to install without the AirPod test decoder.

https://docs.openmqttgateway.com/dev/upload/web-install.html

All recent Decoder changes will be in the next release and subsequently in the next upcoming OpenMQTTGtateway release.

**Reply by andrewplas** [reply]
Amazing, thank you @DigiH 
I was using the Olimex eth-poe-iso board using the web installer without issue, but was playing with a custom _env.ini in PlatformIO to connect some DS18B20 temp sensors.  I made that change in the platformio.ini file, uploaded, and the SBHT003C is showing up and working great!  Thank you so much!

**Comment by chrestme** [comment]
I don’t have any thing else at this time. Thanks for your help with this.
